openedx_tagging/core/tagging/api.py

```diff
@@ -100,5 +116,180 @@ def tag_object(
+            if csv_fields != header_fields:
+                raise ValueError(
+                    _(
+                        f"Invalid CSV header: {header_fields}. Must be: {csv_fields}."
+                    )
+                )
+            tags_data = list(csv_reader)
+        elif format == TaxonomyDataFormat.JSON:
+            tags_data = json.load(tags)
+            if 'tags' not in tags_data:
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing 'tags' list."
+                    )
+                )
+            tags_data = tags_data.get('tags')
+        else:
+            raise ValueError(
+                _(
+                    f"Invalid format: {format}"
+                )
+            )
+    except ValueError as e:
+        raise e
+    finally:
+        tags.close()
+
+
+    new_tags = []
+    updated_tags = []
+
+    def create_update_tag(tag):
+        """
+        Function to create a new Tag or update an existing one.
+
+        This function keeps a creation/update history with `new_tags` and `updated_tags`,
+        a same tag can't be created/updated in a same taxonomy import.
+        Also, recursively, creates the parents of the `tag`.
+
+        Returns the created/updated Tag.
+        Raise KeyError if 'id' or 'name' don't exist on `tag`
+        """
+
+        tag_id = tag['id']
+        tag_name = tag['name']
+        tag_parent_id = tag.get('parent_id')
+        tag_parent_name = tag.get('parent_name')
+
+        # Check if the tag has not already been created or updated
+        if tag_id not in new_tags and tag_id not in updated_tags:
+            try:
+                # Update tag
+                tag_instance = taxonomy.tag_set.get(external_id=tag_id)
+                tag_instance.value = tag_name
+
+                if tag_instance.parent and (not tag_parent_id or not tag_parent_name):
+                    # if there is no parent in the data import
+                    tag_instance.parent = None
+                updated_tags.append(tag_id)
+            except Tag.DoesNotExist:
+                # Create tag
+                tag_instance = Tag(
+                    taxonomy=taxonomy,
+                    value=tag_name,
+                    external_id=tag_id,    
+                )
+                new_tags.append(tag_id)
+
+            if tag_parent_id and tag_parent_name:
+                # Parent creation/update
+                parent = create_update_tag({'id': tag_parent_id, 'name': tag_parent_name})
+                tag_instance.parent = parent
+
+            tag_instance.save()
+            return tag_instance
+        else:
+            # Returns the created/updated tag from history
+            return taxonomy.tag_set.get(external_id=tag_id)
+
+    # Create and update tags
+    with transaction.atomic():
+        # Delete all old Tags linked to the taxonomy
+        if replace:
+            Tag.objects.filter(taxonomy=taxonomy).delete()
```

Comment:
Instead of deleting all the existing Tags and recreating them, can you do something like what `tag_object` does?

* [Grab the taxonomy's current list of tags](https://github.com/openedx/openedx-learning/pull/57/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R241-R247), and map them by `external_id`.
* [pop existing tags off the list](https://github.com/openedx/openedx-learning/pull/57/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R250) before updating them, or add new ones if they're not in the current list.
* if replace: [delete any tags that weren't popped off](https://github.com/openedx/openedx-learning/pull/57/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R280).

Otherwise, tag name changes won't get reflected correctly on existing ObjectTags, because:

* object_tag._name = "the old name"
* object_tag.tag is now None, so
* ObjectTag.resync() doesn't have a Tag reference to check for an updated name.

To verify this happens as expected, could you add a test case?

Reply:
@pomegranited Thanks. I used the [updated_tags](https://github.com/open-craft/openedx-learning/pull/1/files#diff-951a5c4d32e1301ac1c30419ebce4ada8ae17db6c93f4c2a5a34caac9eb30e0aR244) to do this. Also I added a [test case](https://github.com/open-craft/openedx-learning/pull/1/files#diff-6cda6f74e9f6bf053923d015688ea76e50cfec92ef244df48aaa3942288fcbb5R314)

Reply:
That looks perfect, thank you @ChrisChV !